### DISCUSSION TITLE
Disable an Rspamd rule in local.d/force_actions.conf

### DISCUSSION THREAD

**Discussion by manuthhsl** [Mailu/Mailu#3393] [User support]
I'm trying to set up Mailu for a setup that involves receiving emails with the sender address or domain that's different from SMTP username. I'm hitting a roadblock right now - Mailu rejects the emails even though I've added the domain in question to SPF and DKIM whitelist maps, and indeed Rspamd history page says that SPF and DKIM whitelists are applied but the mails are rejected because of the rule ANTISPOOF_DMARC_ENFORCE_LOCAL in the file local.d/force_actions.conf.

![Screenshot 2024-08-31 alle 14 39 43](https://github.com/user-attachments/assets/027cac02-e57a-4d4c-9a5a-c6d4bb184313)

What I've tried so far:
- mount a file from Docker host without this rule to replace the built-in /etc/rspamd/local.d/force_actions.conf: setting this mount to read-only leads to the container to fail to start (it doesn't like read-only permission), setting it to read-write will make rspamd put this rule back in
- creating the file overrides/rspamd/force_actions.conf in the Docker host either without this rule or changing 'action' to 'accept': no change to the container's /etc/rspamd/local.d/force_actions.conf
- mounting this custom force_actions.conf to /etc/rspamd/override.d: also no change to the container's force_actions.conf file

I would like to know how to disable this one rule, thanks.

**Comment by manuthhsl** [comment]
OK I've think I've found how: 

- Create a file with this content: `rspamd_config:add_condition('ANTISPOOF_DMARC_ENFORCE_LOCAL', function(task) return false end)`
- Mount it to /etc/rspamd/rspamd.local.lua
- Create a file with this content:
```
rules {
  ANTISPOOF_NOAUTH {
    action = "reject";
    expression = "!IS_LOCALLY_GENERATED & !MAILLIST & ((IS_LOCAL_DOMAIN_E & MISSING_FROM) | (IS_LOCAL_DOMAIN_H & (R_DKIM_NA & R_SPF_NA & DMARC_NA & ARC_NA)))";
    message = "Rejected (anti-spoofing: noauth). Please setup DMARC with DKIM or SPF if you want to send emails from your domain from other servers.";
  }
  ANTISPOOF_DMARC_ENFORCE_LOCAL {
    action = "no action";
    expression = "!IS_LOCALLY_GENERATED & !MAILLIST & (IS_LOCAL_DOMAIN_H | IS_LOCAL_DOMAIN_E) & (DMARC_POLICY_SOFTFAIL | DMARC_POLICY_REJECT | DMARC_POLICY_QUARANTINE | DMARC_NA)";
    message = "Rejected (anti-spoofing: DMARC compliance is enforced for local domains, regardless of the policy setting)";
  }
  ANTISPOOF_AUTH_FAILED {
    action = "reject";
    expression = "!IS_LOCALLY_GENERATED & !MAILLIST & BLACKLIST_ANTISPOOF";
    message = "Rejected (anti-spoofing: auth-failed)";
  }
  ANTIVIRUS_FLAGGED {
    action = "reject";
    expression = "CLAM_VIRUS | OLETOOLS_MACRO_MRAPTOR | OLETOOLS_MACRO_SUSPICIOUS";
    message = "Rejected (dangerous/malicious code detected)";
  }
  ANTIVIRUS_FAILED {
    action = "soft reject";
    expression = "CLAM_VIRUS_FAIL | OLETOOLS_FAIL";
    message = "Please retry later (anti-virus/oletools not ready)";
  }
}
```
- Mount it to override.d/force_actions.conf

**Reply by TomFyuri** [reply]
Can you run this command in your docker antispam container?
`rspamadm configdump force_actions`

I've tried your steps on the recent mailu version and it doesn't work for me.

edit: OK. I've done similar steps in the end but instead force_actions.conf on my end is like this:
```
rules {
  ANTISPOOF_DMARC_ENFORCE_LOCAL {
    action = "no action";
  }
}
```
while `rspamadm configdump force_actions` shows ANTISPOOF_DMARC_ENFORCE_LOCAL rule with both actions reject and no action existing at the same time - yet it doesn't seem to trigger anymore.

**Comment by fracture-point** [comment]
Hope this helps for folks - this was an hour of headslamming with things that didn't work, BUT **I got it to work with a clever workaround**.

**Working Solution: Expression Override**
- Override rule `expression` with logically impossible condition
- rspamd's `duplicate=merge` creates config with both expressions:
  ```
  expression = "<original complex condition>";
  expression = "!IS_LOCALLY_GENERATED & IS_LOCALLY_GENERATED";
  ```
- Either rspamd uses "last wins" (uses impossible expression) OR ANDs them together
- Either way, the result is always false → rule never matches → emails delivered

So the output of `docker exec mailu-antispam-1 rspamadm configdump force_actions 2>/dev/null | grep -A8 'ANTISPOOF_DMARC_ENFORCE_LOCAL'`
...looks like...

```
    ANTISPOOF_DMARC_ENFORCE_LOCAL {
        message = "Rejected (anti-spoofing: DMARC compliance is enforced for local domains, regardless of the policy setting)";
        expression = "!IS_LOCALLY_GENERATED & !MAILLIST & (IS_LOCAL_DOMAIN_H | IS_LOCAL_DOMAIN_E) & (DMARC_POLICY_SOFTFAIL | DMARC_POLICY_REJECT | DMARC_POLICY_QUARANTINE | DMARC_NA)";
        expression = "!IS_LOCALLY_GENERATED & IS_LOCALLY_GENERATED";
        action = "reject";
    }
    ANTIVIRUS_FAILED {
        message = "Please retry later (anti-virus/oletools not ready)";
        expression = "CLAM_VIRUS_FAIL | OLETOOLS_FAIL";
```
## Config
My config is all inline on the compose file so I can manage it better in source...

```
services:
  antispam:
    image: ${DOCKER_ORG:-ghcr.io/mailu}/${DOCKER_PREFIX:-}rspamd:${MAILU_VERSION:-2024.06}
   ...
    configs:
      - source: rspamd_dmarc_config
        target: /overrides/dmarc.conf
        mode: 0444
      - source: rspamd_force_actions_disable
        target: /overrides/force_actions.conf
        mode: 0444

...

configs:
  rspamd_dmarc_config:
    content: |
      # DMARC Configuration Override
      # Relax strict enforcement to respect published DMARC policy levels
      # instead of rejecting ALL DMARC failures regardless of policy

      # Respect the sender's published DMARC policy:
      # - p=reject -> reject the message
      # - p=quarantine -> add header and deliver to spam (was: reject)
      # - p=none -> add header only (monitoring)

      actions {
        reject = 15.0;
        quarantine = "add header";
        softfail = "add header";
      }

  # Rspamd force_actions configuration - disable local domain enforcement
  rspamd_force_actions_disable:
    content: |
      # Disable specific force_actions rules by overriding their expressions
      # to conditions that will never match (always false)
      # This works around rspamd's duplicate=merge behavior

      rules {
        ANTISPOOF_DMARC_ENFORCE_LOCAL {
          # Override expression to never match - disables the rule
          expression = "!IS_LOCALLY_GENERATED & IS_LOCALLY_GENERATED";
        }
        ANTISPOOF_AUTH_FAILED {
          # Override expression to never match - disables the rule
          expression = "!IS_LOCALLY_GENERATED & IS_LOCALLY_GENERATED";
        }
      }
```

# Failed Approaches 

The following approaches were attempted but did not work due to rspamd's configuration behavior in Mailu 2024.06:

**1. Config-based action override (first attempt)**
- Tried: Override `action = "reject"` to `action = "add header"`
- Failed: rspamd's `duplicate=merge` behavior kept both actions, first one (reject) wins
- Result: Emails still rejected

**2. Lua `add_condition()` in `/overrides/rspamd.local.lua` (mounted to local.d/)**
- Tried: `rspamd_config:add_condition('FORCE_ACTION_*', function(task) return false end)`
- Failed: Conditions added too early, before force_actions module registered the symbols
- Result: Conditions never applied, emails still rejected

**3. Lua `add_condition()` with `register_finish_script()`**
- Tried: Wrap conditions in `register_finish_script()` callback to ensure symbols exist first
- Failed: force_actions module uses pre-result mechanism which bypasses normal symbol conditions
- Result: Symbols fired despite conditions returning false, emails still rejected

**4. Lua `add_condition()` mounted to `/etc/rspamd/rspamd.local.lua`**
- Tried: Mount Lua script directly to rspamd's root config directory (per Mailu discussion #3393)
- Failed: Same issue - pre-result mechanism bypasses add_condition()
- Result: Emails still rejected

**5. Config-based `enabled = false`**
- Tried: Override rules with `enabled = false;` parameter
- Failed: rspamd doesn't support `enabled` parameter for force_actions rules
- Result: Configuration ignored, emails still rejected

**Reply by nextgens** [reply]
You are tackling this by the wrong end. Ensure the sender of those emails is DMARC authorised (so get it to either DKIM sign or be in the SPF records) and Mailu won't bother you.

**Reply by fracture-point** [reply]
@nextgens - I understand this is not standard - In this case, the domain is used by a distributed organization with divisions run completely separately of each other, where there is zero chance of getting everyone to set up their various email systems properly. Strict enforcement of a "local" domain assumes that you have complete and total control over the way the domain is used, which is a poor assumption for the zillions of distributed, non-profit, volunteer etc. organizations out there. I think a safety-first posture is fine, but an override parameter should have been added when mailu decided to add hard enforcement (a breaking change in this case).

**Reply by nextgens** [reply]
Mailu's mission statement is quite clear, it's meant to be insular.

The easy fix for organizations you describe is to hand over control of subdomains.
